File: Swift/ExampleTests/common/CartInputString.swift
public struct CartInputString: CustomStringConvertible, Equatable {
    public let totalItems: String
    public let shipping: String
    public let discount: String
    public let totalPrice: String
    public let notes: String

    public init(totalItems: String, shipping: String, discount: String, totalPrice: String, notes: String) {
        self.totalItems = totalItems
        self.shipping = shipping
        self.discount = discount
        self.totalPrice = totalPrice
        self.notes = notes
    }

    public init(fromArray v: [String]) {
        self.totalItems = v.count > 0 ? v[0] : ""
        self.shipping = v.count > 1 ? v[1] : ""
        self.discount = v.count > 2 ? v[2] : ""
        self.totalPrice = v.count > 3 ? v[3] : ""
        self.notes = v.count > 4 ? v[4] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> CartInputString {
        let parts = Tokens.require(text, 5, "CartInput")
        return CartInputString(totalItems: parts[0], shipping: parts[1], discount: parts[2], totalPrice: parts[3], notes: parts[4])
    }

    public var description: String {
        return Tokens.token(totalItems) + " " + Tokens.token(shipping) + " " + Tokens.token(discount) + " " + Tokens.token(totalPrice) + " " + Tokens.token(notes)
    }

    public static let dncString = "?DNC?"

    public static func == (a: CartInputString, b: CartInputString) -> Bool {
        return (a.totalItems == dncString || b.totalItems == dncString || a.totalItems == b.totalItems)
            && (a.shipping == dncString || b.shipping == dncString || a.shipping == b.shipping)
            && (a.discount == dncString || b.discount == dncString || a.discount == b.discount)
            && (a.totalPrice == dncString || b.totalPrice == dncString || a.totalPrice == b.totalPrice)
            && (a.notes == dncString || b.notes == dncString || a.notes == b.notes)
    }
}

File: Python/ExampleTests/common/cart_input_string.py
from . import tokens
DNC_STRING = '?DNC?'


class CartInputString:
    def __init__(self, total_items: str = '', shipping: str = '', discount: str = '', total_price: str = '', notes: str = ''):
        self.total_items = total_items
        self.shipping = shipping
        self.discount = discount
        self.total_price = total_price
        self.notes = notes

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else '',
            v[4] if len(v) > 4 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 5, 'CartInput')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3],
            parts[4]
        )

    def __str__(self):
        return (tokens.token(self.total_items) + ' ' +
                tokens.token(self.shipping) + ' ' +
                tokens.token(self.discount) + ' ' +
                tokens.token(self.total_price) + ' ' +
                tokens.token(self.notes))

    def _key(self):
        return (self.total_items, self.shipping, self.discount, self.total_price, self.notes)

    def __eq__(self, other):
        if not isinstance(other, CartInputString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'CartInputString({self})'
